pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,442 @@
+				VirtualClusterName: install.SystemInterfaceName,
+			})
+			if err != nil {
+				return err
+			}
+
+			pgURL, err := copyPGCertsAndMakeURL(ctx, cm.t, cm.c, srcNode, srcClusterSetting.PGUrlCertsDir, addr[0])
+			if err != nil {
+				return err
+			}
+
+			sourceChan <- sourceTenantInfo{name: h.Tenant.Descriptor.Name, pgurl: pgURL}
+
+			l.Printf("waiting for destination tenant to be created")
+			chanReadCtx(ctx, cm.destStartedChan)
+			l.Printf("done")
+
+			return nil
+		},
+	)
+
+	cm.destMvt.OnStartup("create destination tenant on standby",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			l.Printf("waiting to hear from source cluster")
+			sourceInfo := chanReadCtx(ctx, sourceChan)
+
+			if err := h.Exec(r, fmt.Sprintf(
+				"CREATE TENANT %q FROM REPLICATION OF %q ON '%s'",
+				destTenantName, sourceInfo.name, sourceInfo.pgurl.String(),
+			)); err != nil {
+				return errors.Wrap(err, "creating destination tenant")
+			}
+
+			if err := h.QueryRow(r, fmt.Sprintf(
+				"SELECT job_id FROM [SHOW JOBS] WHERE job_type = '%s'",
+				replicationJobType,
+			)).Scan(&cm.ingestionJobID); err != nil {
+				return errors.Wrap(err, "querying ingestion job ID")
+			}
+
+			l.Printf("replication job: %d. Let initial scan complete", cm.ingestionJobID)
+			// TODO(msbutler): relax requirement that initial scan completes before upgrades.
+			if err := cm.WaitForReplicatedTime(ctx, timeutil.Now(), h, r, 5*time.Minute); err != nil {
+				return err
+			}
+			close(cm.destStartedChan)
+			return nil
+		})
+}
+
+func (cm *c2cMixed) WorkloadHook(ctx context.Context) {
+	tpccInitCmd := roachtestutil.NewCommand("./cockroach workload init tpcc").
+		Arg("{pgurl%s}", cm.c.Range(1, cm.sp.srcNodes)).
+		Flag("warehouses", 10)
+	tpccRunCmd := roachtestutil.NewCommand("./cockroach workload run tpcc").
+		Arg("{pgurl%s}", cm.c.Range(1, cm.sp.srcNodes)).
+		Option("tolerate-errors").
+		Flag("warehouses", 100)
+	cm.workloadStopper = cm.sourceMvt.Workload("tpcc", cm.c.WorkloadNode(), tpccInitCmd, tpccRunCmd)
+}
+
+func (cm *c2cMixed) LatencyHook(ctx context.Context) {
+	cm.destMvt.BackgroundFunc("latency verifier", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		lv := makeLatencyVerifier("stream-ingestion", 0, cm.sp.maxAcceptedLatency, l,
+			getStreamIngestionJobInfo, nil, true)
+		defer lv.maybeLogLatencyHist()
+		_, db := h.RandomDB(r)
+
+		// The latency verify doesn't need a stopper, as ctx cancellation will stop it.
+		dummyCh := make(chan struct{})
+		if err := lv.pollLatencyUntilJobSucceeds(ctx, db, int(cm.ingestionJobID), time.Second*5, dummyCh); ctx.Err() == nil {
+			// The ctx is cancelled when the background func is successfully stopped,
+			// therefore, don't return a context cancellation error.
+			l.Printf("latency verifier error: %s", err)
+			return err
+		}
+		return nil
+	})
+}
+
+func (cm *c2cMixed) UpdateHook(ctx context.Context) {
```

Comment:
I think a function level comment explaining our general strategy here might be helpful.

Reply:
Done.